src/authorization.ts

```diff
@@ -181,16 +191,81 @@ export class SpendingCondition extends Deserializable {
     nextSigHash: string;
   } {
     const sigHashPreSign = this.makeSigHashPreSign(curSigHash, authType, fee, nonce);
+
     const signature = signWithKey(privateKey, sigHashPreSign);
     const publicKey = getPublicKey(privateKey);
-    const nextSigHash = this.makeSigHashPostSign(sigHashPreSign, publicKey, signature);
+    const publicKeyEncoding = isCompressed(publicKey)
+      ? PubKeyEncoding.Compressed
+      : PubKeyEncoding.Uncompressed;
+    const nextSigHash = this.makeSigHashPostSign(sigHashPreSign, publicKeyEncoding, signature);
 
     return {
       nextSig: signature,
       nextSigHash,
     };
   }
 
+  static nextVerification(
+    initialSigHash: string,
+    authType: AuthType,
+    fee: BigNum,
+    nonce: BigNum,
+    pubKeyEncoding: PubKeyEncoding,
+    signature: MessageSignature
+  ) {
+    const sigHashPreSign = this.makeSigHashPreSign(initialSigHash, authType, fee, nonce);
+
+    const publicKey = createStacksPublicKey(publicKeyFromSignature(sigHashPreSign, signature));
+
+    const nextSigHash = this.makeSigHashPostSign(
+      sigHashPreSign,
+      PubKeyEncoding.Compressed,
+      signature
+    );
+
+    return {
+      pubKey: publicKey,
+      nextSigHash,
+    };
+  }
+
+  static newInitialSigHash(): SpendingCondition {
+    const spendingCondition = new this(
+      AddressHashMode.SerializeP2PKH,
+      '',
+      new BigNum(0),
+      new BigNum(0)
+    );
+    spendingCondition.signerAddress = createEmptyAddress();
+    spendingCondition.pubKeyEncoding = PubKeyEncoding.Compressed;
+    spendingCondition.signature = MessageSignature.empty();
+    return spendingCondition;
+  }
+
+  verify(initialSigHash: string, authType: AuthType): string {
+    if (this.singleSig()) {
+      return this.verifySingleSig(initialSigHash, authType);
+    } else {
+      // TODO: verify multisig
+      return '';
+    }
+  }
+
+  verifySingleSig(initialSigHash: string, authType: AuthType): string {
+    const { pubKey, nextSigHash } = SpendingCondition.nextVerification(
+      initialSigHash,
+      authType,
+      this.fee!,
```

Comment:
Lots of non-null assertions in this PR. Are all these usages in call graphs where these values are guaranteed to be non-null? It's hard to tell when following the function calls.

I think a good solution here would be to find the entry point where these objects with optional properties are provided, perform null checks (throw useful errors where null), then cast using a utility type like `DeepRequired<T>` for use with functions like this one.

Reply:
Yes, fee and nonce values should not be null in all cases. The transaction builders at the top of the call graph will fetch these values if not provided. That said, I can add more null checks in case they're called from other functions later.